File: src/main/java/org/fountainmc/api/world/tileentity/TileEntity.java
package org.fountainmc.api.world.tileentity;

import org.fountainmc.api.BlockType;
import org.fountainmc.api.world.BlockPosition;
import org.fountainmc.api.world.World;
import org.fountainmc.api.world.block.BlockState;

import static com.google.common.base.Preconditions.*;

/**
 * A tile entity in the world, which is extra data associated with a block beyond the type id and BlockState
 */
public interface TileEntity {

    int getX();

    int getY();

    int getZ();

    default BlockPosition getPosition() {
        return new BlockPosition(getWorld(), getX(), getY(), getZ());
    }

    World getWorld();

    BlockType getBlockType();

    /**
     * Returns if the tile entity is still valid.
     * <p>
     * Tile entities will be invalidated if the block changes.
     * Tile entities returned by {@link World#getTileEntity(int, int, int)} should never be invalid, but if you hold on to a reference for too long it might be.
     * </p>
     *
     * @return if valid
     */
    boolean isValid();

    /**
     * Get the blockstate of the tile entity
     * <p>Many tile entities also store data in their blockstates.</p>
     *
     * @return the blockstate
     */
    default BlockState getBlockState() {
        checkState(isValid());
        return getWorld().getBlock(getPosition());
    }

    /**
     * Set the blockstate of the tile entity
     * <p>Many tile entities also store data in their blockstates.</p>
     *
     * @param blockState the blockstate
     */
    default void setBlockState(BlockState blockState) {
        checkArgument(checkNotNull(blockState, "Null BlockState").getBlockType().equals(getBlockType()), "BlockState's type %s doesn't match TileEntity's type %s", blockState, this);
        checkState(isValid());
        getWorld().setBlock(getPosition(), blockState);
    }
}

File: src/main/kotlin/org/fountainmc/api/world/tileentity/TileEntity.kt
package org.fountainmc.api.world.tileentity

import org.fountainmc.api.BlockType
import org.fountainmc.api.world.BlockPosition
import org.fountainmc.api.world.World
import org.fountainmc.api.world.block.BlockState
import kotlin.reflect.KClass
import kotlin.reflect.full.cast

/**
 * A tile entity in the world, which is extra data associated with a block beyond the type id and BlockState
 */
interface TileEntity<S: BlockState> {

    val x: Int

    val y: Int

    val z: Int

    val position: BlockPosition
        get() = BlockPosition(world, x, y, z)

    val world: World

    val blockType: BlockType<S>

    /**
     * Returns if the tile entity is still valid.
     *
     * Tile entities will be invalidated if the block changes.
     * Tile entities returned by [World.getTileEntity] should never be invalid at first,
     * but if you hold on to a reference for too long it might be.
     *
     * @return if valid
     */
    val isValid: Boolean

    /**
     * The blockstate of the tile entity
     *
     * Many tile entities also store data in their blockstates.
     */
    var blockState: S
        get() {
            check(isValid)
            return blockType.stateType.cast(world.getBlock(position))
        }
        set(blockState) {
            require(blockState.blockType == this.blockType) {
                "BlockState's type ${blockState.blockType} doesn't match TileEntity's type ${this.blockType}"
            }
            check(isValid)
            world.setBlock(position, blockState)
        }
}
